components/sync15/src/telemetry.rs

```diff
@@ -0,0 +1,645 @@
+            &Event::new("Object", "Method").value("Value"),
+            json!({"object": "Object", "method": "Method", "value": "Value"}),
+        );
+
+        assert_json(
+            &Event::new("Object", "Method").extra("one", "one".to_string()),
+            json!({"object": "Object",
+             "method": "Method",
+             "extra": {"one": "one"}
+            }),
+        )
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// A Sync failure.
+//
+#[derive(Debug, Serialize)]
+#[serde(tag = "name")]
+pub enum SyncFailure {
+    #[serde(rename = "shutdownerror")]
+    Shutdown,
+
+    #[serde(rename = "othererror")]
+    Other { error: String },
+
+    #[serde(rename = "unexpectederror")]
+    Unexpected { error: String },
+
+    #[serde(rename = "autherror")]
+    Auth { from: String },
+
+    #[serde(rename = "httperror")]
+    Http { code: u32 },
+
+    #[serde(rename = "nserror")]
+    Nserror { code: i32 }, // probably doesn't really make sense in rust, but here we are...
+}
+
+pub fn sync_failure_from_error(e: &Error) -> SyncFailure {
+    SyncFailure::Unexpected {
+        error: e.to_string(),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn reprs() {
+        assert_json(&SyncFailure::Shutdown, json!({"name": "shutdownerror"}));
+
+        assert_json(
+            &SyncFailure::Other {
+                error: "dunno".to_string(),
+            },
+            json!({"name": "othererror", "error": "dunno"}),
+        );
+
+        assert_json(
+            &SyncFailure::Unexpected {
+                error: "dunno".to_string(),
+            },
+            json!({"name": "unexpectederror", "error": "dunno"}),
+        );
+
+        assert_json(
+            &SyncFailure::Auth {
+                from: "FxA".to_string(),
+            },
+            json!({"name": "autherror", "from": "FxA"}),
+        );
+
+        assert_json(
+            &SyncFailure::Http { code: 500 },
+            json!({"name": "httperror", "code": 500}),
+        );
+
+        assert_json(
+            &SyncFailure::Nserror { code: -1 },
+            json!({"name": "nserror", "code": -1}),
+        );
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// Incoming and Outgoing records for an engine's sync
+//
+#[derive(Debug, Serialize)]
+pub struct EngineIncoming {
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    applied: u32,
+
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    failed: u32,
+
+    #[serde(rename = "newFailed")]
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    new_failed: u32,
+
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    reconciled: u32,
+}
+
+impl EngineIncoming {
+    pub fn new() -> Self {
+        Self {
+            applied: 0,
+            failed: 0,
+            new_failed: 0,
+            reconciled: 0,
+        }
+    }
+
+    fn is_empty(inc: &Option<Self>) -> bool {
```

Comment:
This function is a little weird, add a comment saying it's so that wee can use it in a `skip_serializing_if`.